lib/ZwaveClient.ts

```diff
@@ -961,6 +980,173 @@ class ZwaveClient extends TypedEventEmitter<ZwaveClientEventCallbacks> {
+			ScheduleEntryLockSlotId)[] = []
+		for (let i = 1; i <= userCodes; i++) {
+			for (let s = 1; s <= numSlots.numWeekDaySlots; s++) {
+				const slot: ScheduleEntryLockSlotId = {
+					userId: i,
+					slotId: s,
+				}
+				const schedule = await zwaveNode.commandClasses[
+					'Schedule Entry Lock'
+				].getWeekDaySchedule(slot)
+				if (schedule) weeklySchedules.push({ ...slot, ...schedule })
+			}
+
+			for (let s = 1; s <= numSlots.numYearDaySlots; s++) {
+				const slot: ScheduleEntryLockSlotId = {
+					userId: i,
+					slotId: s,
+				}
+				const schedule = await zwaveNode.commandClasses[
+					'Schedule Entry Lock'
+				].getYearDaySchedule(slot)
+				if (schedule) yearlySchedules.push({ ...slot, ...schedule })
+			}
+
+			for (let s = 1; s <= numSlots.numDailyRepeatingSlots; s++) {
+				const slot: ScheduleEntryLockSlotId = {
+					userId: i,
+					slotId: s,
+				}
+				const schedule = await zwaveNode.commandClasses[
+					'Schedule Entry Lock'
+				].getDailyRepeatingSchedule(slot)
+				if (schedule) dailySchedules.push({ ...slot, ...schedule })
+			}
+		}
+
+		const node = this._nodes.get(nodeId)
+		node.schedule = {
+			daily: {
+				numSlots: numSlots.numDailyRepeatingSlots,
+				slots: dailySchedules,
+			},
+			weekly: {
+				numSlots: numSlots.numWeekDaySlots,
+				slots: weeklySchedules,
+			},
+			yearly: {
+				numSlots: numSlots.numYearDaySlots,
+				slots: yearlySchedules,
+			},
+		}
+
+		this.emitNodeUpdate(node, {
+			schedule: node.schedule,
+		})
+
+		return node.schedule
+	}
+
+	async setSchedule(
+		nodeId: number,
+		type: 'daily' | 'weekly' | 'yearly',
+		schedule: ScheduleEntryLockSlotId &
+			(
+				| ScheduleEntryLockDailyRepeatingSchedule
+				| ScheduleEntryLockWeekDaySchedule
+				| ScheduleEntryLockYearDaySchedule
+			)
+	) {
+		const zwaveNode = this.getNode(nodeId)
+
+		if (!zwaveNode?.commandClasses['Schedule Entry Lock'].isSupported()) {
+			throw new Error(
+				'Schedule Entry Lock CC not supported on node ' + nodeId
+			)
+		}
+
+		const slot: ScheduleEntryLockSlotId = {
+			userId: schedule.userId,
+			slotId: schedule.slotId,
+		}
+
+		delete schedule.userId
+		delete schedule.slotId
+
+		const isDelete = Object.keys(schedule).length === 0
+
+		let result: SupervisionResult
+
+		if (type === 'daily') {
+			result = await zwaveNode.commandClasses[
+				'Schedule Entry Lock'
+			].setDailyRepeatingSchedule(
+				slot,
+				schedule as ScheduleEntryLockDailyRepeatingSchedule
+			)
+		} else if (type === 'weekly') {
+			result = await zwaveNode.commandClasses[
+				'Schedule Entry Lock'
+			].setWeekDaySchedule(
+				slot,
+				schedule as ScheduleEntryLockWeekDaySchedule
+			)
+		} else if (type === 'yearly') {
+			result = await zwaveNode.commandClasses[
+				'Schedule Entry Lock'
+			].setYearDaySchedule(
+				slot,
+				schedule as ScheduleEntryLockYearDaySchedule
+			)
+		} else {
+			throw new Error('Invalid schedule type')
+		}
+
+		if (result.status === SupervisionStatus.Success) {
```

Comment:
This will throw if the command was sent without Supervision. The methods return `SupervisionResult | undefined`.

I have a couple of utility functions exported from `@zwave-js/core` that you can use to decide what to do based on the result:
https://github.com/zwave-js/node-zwave-js/blob/4854b0cf2b2c9bad15ac1b42a2632776e28db4b7/packages/core/src/consts/Transmission.ts#L275-L308

If the command **was** unsupervised, you may want to verify the change by reading the slot again, in case it was not accepted by the lock.

Reply:
In poor words if result is undefined I should read the slot again

Reply:
If I don't remember correctly @AlCalzone, the DanaLock might (will?) return false even though codes are set.

Not that I expect a workaround/hack for it, but it might help you to know during development and debigging, that you should pull and check what is set or not and not wasting time on buggy software. 

And there should possible be a refresh button in the UI? 😅

Reply:
In latest commits I added a check and if result is undefined I read the slot back and compare it with the actual set, if they match I return success

Reply:
But I actually think the DanaLock returns `false`

Reply:
you mean a Supervision status of Fail?

Reply:
I quote you're email: 

```
Setting a schedule always results in a supervision status of Fail (0x02). Here's an example for a daily repeating schedule (Mon, Wed, Fri, from 00:00 to 08:30) for user 17, slot 1:

2022-09-01T07:59:44.109Z DRIVER » [Node 002] [REQ] [SendData]

                                  │ transmit options: 0x25

                                  │ callback id:      114

                                  └─[Security2CCMessageEncapsulation]

                                    │ sequence number: 224

                                    └─[SupervisionCCGet]

                                      │ session id:      52

                                      │ request updates: true

                                      └─[ScheduleEntryLockCCDailyRepeatingScheduleSet]

                                          payload: 0x0111012a0000081e

...

2022-09-01T07:59:46.695Z DRIVER « [Node 002] [REQ] [ApplicationCommand]

                                  └─[Security2CCMessageEncapsulation]

                                    │ sequence number: 236

                                    └─[SupervisionCCReport]

                                        session id:          52

                                        more updates follow: true

                                        status:              Fail

                                        duration:            0s

but reading the schedule back afterwards shows that the schedule was indeed accepted by the lock:

2022-09-01T08:00:02.736Z DRIVER » [Node 002] [REQ] [SendData]

                                  │ transmit options: 0x25

                                  │ callback id:      117

                                  └─[Security2CCMessageEncapsulation]

                                    │ sequence number: 227

                                    └─[ScheduleEntryLockCCDailyRepeatingScheduleGet]

                                        payload: 0x1101

...

2022-09-01T08:00:02.952Z DRIVER « [Node 002] [REQ] [ApplicationCommand]

                                  └─[Security2CCMessageEncapsulation]

                                    │ sequence number: 239

                                    └─[ScheduleEntryLockCCDailyRepeatingScheduleReport]

                                        payload: 0x11012a0000081e

 

 

In that case, I would have expected a supervision status of Success (0xff) instead, to signal to the controlling node that the schedule was accepted.
```